Adding Audible Status Notification (#864)
smartdevicelink/sdl_evolution: proposals/NNNN-Adding-Audible-Status-Notification.md

# Adding Audible Status Notification

* Proposal: [SDL-NNNN](NNNN-Adding-Audible-Status-Notification.md)
* Author: [Shohei Kawano](https://github.com/Shohei-Kawano)
* Status: **Create**
* Impacted Platforms: [Core / iOS / Java Suite / RPC /HMI]

## Introduction

Currenlty, there is no way to independently notify the Audible status to applications that output audio such as SDL Media.  
This time, `OnAudibleStatus` is added as a new API so that the status of Audible can be notified.

## Motivation

> As a means to notify the Audible status to applications that output audio.  

Reviewer: I'm not sure what you mean here.  Does it make sense to update to "The motivation for this proposal is to allow applications that output audio the ability to get an Audible status." ?

Author: No, This proposal is what HMI wants to control AudibleStatus independently.
So, referring to yours, I think that it is better to do as follows.
"The motivation for this proposal is to allow the ability to set an Audible status to applications that output audio."

Reviewer: Okay, please update to ""The motivation for this proposal is to allow the ability to set an Audible status to applications that output audio."

Author: commit 629e7a12616e82ae2dbc4447e5baa78256489583
Author: 河野 祥平 <103082@ad.nexty-ele.com>
Date:   Thu Nov 14 12:26:50 2019 +0900

    Modify description of Motivatioi.

> In the case of Mobile from Core, Audible without changing the HMI Level by using the parameters `audioStreamingState` and hmiLevel in RPC `OnHMIStatus` Can be notified.  

Reviewer: I'm not sure what you mean by "of Mobile from Core" - can you please clarify?

Author: It mean "Notification from Mobiel to Core in MOBILE_API"
Is this clarify?

Reviewer: Are you suggesting that mobile requests an audio state, or do you want to use the hmi to update the audio state without changing the hmi level?

Reviewer: If we understand your request correctly, we'd suggest revising to the following:
```suggestion
Currently, the `OnHMIStatus` notification from Mobile to Core can set an Audible state without changing the hmiLevel by using the parameters `audioStreamingState` and hmiLevel
```

> However, in the case of Core from HMI, there is no appropriate API to notify the Audible status by itself,  

Reviewer: Same question as above - can you please provide clarification regarding "Core from HMI" here?

Author: It mean "Notification from HMI to Core in HMI_API"

Reviewer: ```suggestion
However, there is currently no notification from HMI to Core to set the Audible status,
```

> so a new API to notify the audio status can be added to notify the audio status regardless of the screen status Like that.  

Reviewer: ```suggestion
so a new API to notify the audio status can be added regardless of the screen status.  
```

Reviewer: This doesn't appear to be resolved

Author: committed

As a specific example, there is a specification that allows a user to display and receive an incoming call with Notification, Overlay or ONS without changing to the telephone screen when an incoming call is received.  
In this case, since there is no appropriate API to notify the Audible status independently, using `OnEventChanged (PHONE_CALL)` displays Music App on the screen, but the HMI Level becomes Background.  
As a response method, we propose to add a new API to notify the Audible status. 

## Proposed solution

A new API `OnAudibleStatus` that can be used with the HMI API is added.  
API parameters to be added are `appName` and `audibleState`, so that the target application and Audible status can be notified.  

HMI API  

> ```

Reviewer: Can you please add the language used here so that the code block formats properly (for example: ```xml)

Author: Add the code block formats properly and commited it.
====
commit 1cd112922ce6588717143fcf2b119f01127cad27 (HEAD -> master, origin/master, origin/HEAD)
Author: 河野 祥平 <103082@ad.nexty-ele.com>
Date:   Wed Nov 13 08:39:26 2019 +0900

    the code block formats properly
====

<function name="OnAudibleStatus" messagetype="notification">	
    <description>	
        It can notify the Audible status independently of the screen status.	
    </description>	
    <param name="appName" type="String" maxlength="100" mandatory="true">	
        <description> Application Name </description>	
    </param>	
    <param name="audibleState" type="AudibleState" defvalue="AUDIBLE" mandatory="true">	

>         <description> See AudibleState. If it is NOT_AUDIBLE, the app must stop voice to SDL Core(stop service). </description>	

Reviewer: ```suggestion
        <description> See AudibleState. If it is NOT_AUDIBLE, the app must stop voice to SDL Core (stop service). </description>	
```

    </param>	
</function>	
	
	
<enum name="AudibleState">	
    <description> Enumeration that describes possible states of audio output. </description>	
    <element name="AUDIBLE"/>	
    <element name="ATTENUATED"/>	
    <element name="NOT_AUDIBLE"/>	
</enum>	
```

## Potential downsides

Nothing.

## Impact on existing code

> No Impact. Because, just add new API and enum.

Reviewer: Can you please specify the changes required for the platforms you noted as impacted: Core / iOS / Java Suite / RPC /HMI?

Author: Core and HMI need to change.
So, I modified it.
====
commit d2ee1a94caba2f00967f042082853180dbae17ae (HEAD -> master, origin/master, origin/HEAD)
Author: 河野 祥平 <103082@ad.nexty-ele.com>
Date:   Wed Nov 13 08:56:01 2019 +0900

    Modify Impact on existing code
====

Reviewer: Please provide the impact on iOS, Java Suite, and the RPC Spec as well, thanks!

Author: commit c76dff738b4c3f78903c9adfb7e2250cd002967b
Author: 河野 祥平 <103082@ad.nexty-ele.com>
Date:   Thu Nov 14 11:46:34 2019 +0900

    Modify description of Impacted platforms

## Alternatives considered

> There is no alternative. 

Reviewer: I think this sentence can be removed, given my suggested change to your next sentence.

Author: If the description is not required, I can remove it.

Reviewer: Yes, as your next sentence now provides an alternative you considered, this sentence can be removed.  Thanks!

Author: commit 57239be4d8848312e3bb286722db114ce8334b9a (HEAD -> master, origin/master, origin/HEAD)
Author: 河野 祥平 <103082@ad.nexty-ele.com>
Date:   Thu Nov 14 12:28:20 2019 +0900

    remove desciption of Alternatives considerd

> There is idea to add only parameters to the existing API. However, there is no existing API that suits the purpose, and since the merit of being able to handle audio alone is losted, it seems not good.

Reviewer: ```suggestion
I considered only adding parameters to the existing API. However, there is no existing API that suits the purpose, and since the merit of being able to handle audio alone is lost, this does not seem to be a good alternative.
```
